pressurecooker/images.py

```diff
@@ -21,50 +25,108 @@
-        while x_index < root and index < len(images):
-            im = ImageOps.fit(images[index], (int(offset), int(offset)), Image.ANTIALIAS)
-            new_im.paste(im, (int(offset * x_index), int(offset * y_index)))
-            x_index += 1
-            index += 1
-        y_index += 1
-    new_im.save(fpath_out)
 
-def create_image_from_pdf_page(fpath_in, fpath_out, page_number=0):
+
+# SMARTCROP UTILS
+################################################################################
+
+THUMBNAIL_SIZE = (400, 225)  # 16:9
+
+def scale_and_crop_thumbnail(image, size=THUMBNAIL_SIZE, crop="smart", **kwargs):
+    """
+    Scale and crop the PIL Image ``image`` to maximum dimensions of ``size``.
+    By default, ``crop`` is set to "smart" which will crop the image down to size
+    based on the entropy content of the pixels. The other options are:
+    * Use ``crop="0,0"`` to crop from the left and top edges
+    * Use ``crop=",0"`` to crop from the top edge.
+    Optional keyword arguments:
+    * ``zoom=X``: crop outer X% before starting
+    * ``target``: recenter here before cropping (default center ``(50, 50)``)
+    See the ``scale_and_crop`` docs in ``thumbscropping.py`` for more details.
+    """
+    return scale_and_crop(image, size, crop=crop, upscale=True, **kwargs)
+
+
+
+# THUMBNAILS FOR CONTENT KINDS
+################################################################################
+
+def get_image_from_epub(epubfile, fpath_out, crop=None):
+    book = ebooklib.epub.read_epub(epubfile)
+    # 1. try to get cover image from book metadata (content.opf)
+    cover_item = None
+    covers = book.get_metadata('http://www.idpf.org/2007/opf', 'cover')
+    if covers:
+        cover_tuple = covers[0] # ~= (None, {'name':'cover', 'content':'item1'})
+        assert cover_tuple[1]['name'] == 'cover', 'wrong key name'
+        cover_item_id = cover_tuple[1]['content']
+        for item in book.items:
+            if item.id == cover_item_id:
+                cover_item = item
+    if cover_item:
+        image_data = BytesIO(cover_item.get_content())
+    else:
+        # 2. fallback to get first image in the ePub file
+        images = list(book.get_items_of_type(ebooklib.ITEM_IMAGE))
+        # TODO: get largest image of the bunch
+        if not images:
+            return
+        image_data = BytesIO(images[0].get_content())
+
+    # Save image_data to fpath_out
+    im = Image.open(image_data)
+    im = scale_and_crop_thumbnail(im, crop=crop)
+    im.save(fpath_out)
+
+
+def get_image_from_zip(htmlfile, fpath_out, crop="smart"):
+    biggest_name = None
+    size = 0
+    with zipfile.ZipFile(htmlfile, 'r') as zf:
+        valid_exts = [file_formats.PNG, file_formats.JPEG, file_formats.JPG]
+        valid_files = filter(lambda f: os.path.splitext(f)[1][1:] in valid_exts, zf.namelist())
+        # get the biggest (most pixels) image in the zip.
+        for filename in valid_files:
+            _, ext = os.path.splitext(filename)
```

Comment:
I don't see `ext` used?

Reply:
fixed (by removing the reduntant filter on top)